### ISSUE TITLE
embedding 알고리즘 탐색

### ISSUE THREAD

**Issue by Hyeonjha** [Hyeonjha/AIV_ImageEmbedding#1]
Original comment by JihoonPark93 - Jun 3

- 이미지 크기에 관계 없이 동작할 수 있어야 함
- output vector dimension size 는 제한이 없음
- CPU, GPU 환경 구분 없이 동작할 수 있는 알고리즘 모두 탐색
- 찾아진 알고리즘의 동작 확인까지가 issue 의 목표

**Comment by Hyeonjha** [comment]
Jun 4

https://github.com/MizzleAa/Faiss-Image-Relation/tree/main?tab=readme-ov-file
위 사이트 속 코드 참고 → mobilenet을 train data로 학습시킨 후 test.

단순 단품 이미지 분류 예제를 돌려본 결과 성능이 나쁘지 않음 
(epoch 100으로 다 돌리기엔 시간이 없어서 10번만 돌렸는데 이때 최대 성능은 85%)

실제 적용시켜야할 데이터로 돌려보니 모델 학습은 되지만 predict에서 segmentation fault 발생. (여기서도 데이터 크기 너무 커서 줄여서 사용)

1. 코드는 그대로 가져다썼기 때문에 예제 data와 다른점 찾아보고 그에 맞게 코드 변경
2. train.py 잘 구현되고 model 파일 생성 오류없이 잘 됐는지 확인
3. model 사이즈가 커서 발생한 경우 batch size 줄여보기

추가 해결방안 시도해보거나 생각하는게 있으면 업데이트하겠습니다!

**Comment by Hyeonjha** [comment]
Jun 4

train data 카테고리 갯수 예제 data와 같게끔 바꿔서 돌리니 에러 해결됐습니다!
train data - 6개의 카테고리, 각 카테고리별 60개의 데이터, epoch = 10, batch_size = 20 으로 모델 학습시켰을때의 결과는 다음과 같습니다.
Epoch 1/10, Loss: 4.2164988339992995, Accuracy: 0.51%
Epoch 2/10, Loss: 1.33896099266253, Accuracy: 0.55%
Epoch 3/10, Loss: 1.2941943072436148, Accuracy: 0.61%
Epoch 4/10, Loss: 1.27920320054941, Accuracy: 0.58%
Epoch 5/10, Loss: 1.2310224860383754, Accuracy: 0.65%
Epoch 6/10, Loss: 1.2074256821682579, Accuracy: 0.61%
Epoch 7/10, Loss: 1.1897772142761631, Accuracy: 0.46%
Epoch 8/10, Loss: 1.1480193775996828, Accuracy: 0.51%
Epoch 9/10, Loss: 1.13383969821428, Accuracy: 0.56%
Epoch 10/10, Loss: 1.2098213716557151, Accuracy: 0.55%

모델 학습을 간단하게만 시켜서 그런지 predict.py 돌린결과 정확도가 떨어지네요.
train data 갯수 늘리고 epoch 늘리면서 모델 학습시키면서 성능 확인해보도록 하겠습니다!

내일 진행하고자 하는 작업에 대한 질문이 있습니다!
1. train data의 카테고리와 갯수가 많아서 모델 학습에 꽤 오랜 시간이 걸리는데 데이터를 조금씩 늘려가면서 돌려봐야할까요?
2. 오늘 사용한 코드를 설명해주신 기능에 맞게 발전시키는 형태로 만들고자 하는데 괜찮을까요?

**Comment by Hyeonjha** [comment]
Jun 4

Original comment by JihoonPark93

가지고 있는 데이터셋으로 학습해서 사용하지 않고, pretrianed weight 를 사용해보면 좋을 것 같아요!
(참고 https://pytorch.org/vision/main/models)
개발하고자 하는 기능은 학습 데이터를 얼마만큼 잘 분류하는지를 평가하기 보다는, 어떤 데이터가 들어올 지 완전히 모르는 상태에서 무작위로 들어오는 이미지 데이터에 대해서 embedding 을 구하고, 그 embedding 이 유사한 이미지에서 유사하게 나오는지를 확인하는 게 더 중요할 것 같아요!

그래서 내일 아래 내용으로 진행해보면 어떨까 싶어요
- class similarity 계산 방식 고민해보기 
    - 어떤 방식으로 동일 class 안에서 similarity 를 평가할 수 있을지?
    - ![image](https://github.com/JihoonPark93/ImageImbedding/assets/65285548/8aaac8f3-13c1-4b4e-9fbb-9ec1d167de37)
        - 위 이미지의 오른쪽 처럼 같은 이미지끼리 잘 분류했는지를 평가하는 방법을 찾는거라고 생각하시면 됩니다!
    - ex. A class 의 이미지와 모든 이미지의 similarity 를 계산했을 때 similarity threshold 를 몇으로 했을 때 동일한 class 가 가장 많이 분류 되고, 다른 class 가 가장 적게 분류되는지?
    - ex. N 개의 이미지를 자신을 제외한 N-1 개의 데이터와 similarity 를 계산했을 때 잘 분류된 class, 잘못 분류된 class 의 개수를 파악하기
    - 이 외에 좋은 평가 방법이 있다면 적용하기
    - 당연히 코드로 구현돼서 평가 결과를 수치로 확인할 수 있으면 좋을 것 같습니다

- pre-train weight 을 사용해 다양한 model 에 대해 저희 data set의 같은 class 끼리 similarity 성능을 평가하기
    - 위에 구현된 평가 방식을 적용
    - 각 model 별 processing time 도 분석
        - 100회 반복 시간을 구해, 평균값으로 분석해주세요

아래 링크에서 이미 다양한 모델에 대해 embedding 을 구하는 코드를 아주 잘 작성해둬서 참고해보면 좋을 것 같아요 ! similiarity 계산하는 부분까지 있네요
https://github.com/christiansafka/img2vec/blob/master/img2vec_pytorch/img_to_vec.py

혹시 이해 안되면 연락주세요!

**Comment by Hyeonjha** [comment]
Jun 5

[ 6/5 정리 ]
오늘 위에 정리해주신 내용들 그리고 깃헙 코드 참고하여 진행했습니다!
(similarity_.ipynb : 오늘 진행한 내용 간결하게 정리한 파일 → 계속 진행
gitCode.ipynb : 오늘 이것저것 돌려본 내용들 전부 넣어놓은 파일 )
1. Nearest Neighbor 기반 평가
2. Threshold 기반 평가
3. Similarity Threshold
4. Leave-One_Out Cross-Similarity 
5. KNN(K-Nearest Neighbors)
6. Confusion Matrix
7. Cluster Analysis
8. Pairwise Similarity Distribution
9. Metric Learning

추가로 similarity 관련 아이디어 정리했고 오늘 1, 2, 3, 4번으로 계속 진행했는데 처음에 코드짜다가 헷갈려서 헤매느라 시간이 좀 걸렸네요..!

일단 pre-trained model을 사용해서 오늘 진행한 결과로는 성능이 그렇게 좋지 않았어요. 좀 더 저희 dataset에 잘맞는 class similarity 계산 방식과 pre-trained 모델을 찾아보겠습니다.

아직 방향이 잡히지 않아서  오늘 하던 작업을 계속 진행해야 할 것 같아요...!
이후에 해야 할 일은 어느정도 정해놔서 다음엔 좀 더 빠르게 진행하도록 하겠습니다!

**Comment by Hyeonjha** [comment]
Jun 5 
Original comment by JihoonPark93

금요일에 작성해주신 코드랑 결과 한번 설명 부탁드려요!

**Comment by Hyeonjha** [comment]
Jun 7

1. Nearest Neighbor 기반 평가
    - 각 이미지의 임베딩 계산
    - 각 이미지에 대해 다른 모든 이미지와의 유사도 계산
    - 유사도 가장 높은 이미지의 label을 예측 레이블로 설정
    - 예측 레이블이 실제 레이블과 일치하는지 확인 → 정확도 계산
2. Threshold 기반 평가
    - 모든 이미지 간의 코사인 유사도 계산
    - 동일한 레이블을 가진 이미지 쌍의 유사도가 주어진 임계값보다 높은지,
    - 다른 레이블을 가진 이미지 쌍의 유사도가 주어진 임계값보다 낮은지 ⇒ 정확도 평가
    - * 각 임계값에서 모델의 정확도 평가, 그래프 시각화
3. Similarity Threshold 방법:
    * A 클래스의 이미지와 모든 이미지의 유사도를 계산한 후, 유사도 임계값(threshold)을 조정하여 동일한 클래스가 가장 많이 분류되고 다른 클래스가 가장 적게 분류되는 임계값을 찾습니다.
    - 모든 이미지 간의 코사인 유사도 계산
    - 다양한 임계값에 대해 유사도가 임계점보다 높은 이미지들 확인, 해당 이미지들의 클래스 조사
    - 동일한 클래스가 가장 많이 분류 & 다른 클래스가 가장 적게 분류되는 임계점 찾기
4. Leave-One-Out Cross-Similarity 방법:
    * N개의 이미지를 각각 자신을 제외한 N-1개의 이미지와 비교하여 동일 클래스와 다른 클래스의 분류 개수를 파악합니다.
5. K-Nearest Neighbors (KNN) 방법:
    * 각 이미지에 대해 k개의 가장 가까운(유사도가 높은) 이미지를 찾습니다.
    * 각 이미지의 k개의 이웃이 동일 클래스인지 확인하여 정확도를 평가합니다.
    * k값을 조정하면서 최적의 k를 찾을 수 있습니다.
    * ** cross-validation 사용 : 모델 평가 시 훈련데이터 사용하지 않도록! **
6. Confusion Matrix 방법:
    * 모든 이미지 쌍에 대해 유사도를 계산하고, 유사도 기준으로 예측된 클래스와 실제 클래스를 비교하여 혼동 행렬(confusion matrix)을 작성합니다.
    * 이 혼동 행렬을 통해 각 클래스별로 얼마나 정확하게 분류되었는지를 분석할 수 있습니다.
7. Cluster Analysis 방법:
    * 모든 이미지의 임베딩을 사용하여 클러스터링 알고리즘(K-means, DBSCAN 등)을 적용합니다.
    * 클러스터링 결과를 기반으로 동일 클래스의 이미지가 얼마나 동일한 클러스터에 속하는지를 평가합니다.
    * 실루엣 점수(Silhouette Score)나 ARI (Adjusted Rand Index)와 같은 지표를 사용할 수 있습니다.
8. Pairwise Similarity Distribution 방법:
    * 동일 클래스 내의 이미지 쌍과 서로 다른 클래스 간의 이미지 쌍에 대해 유사도 분포를 구합니다.
    * 동일 클래스와 다른 클래스의 유사도 분포가 얼마나 분리되는지를 분석합니다. 이는 ROC 곡선(Receiver Operating Characteristic Curve)과 AUC (Area Under the Curve) 값을 통해 평가할 수 있습니다.
((9. Metric Learning 방법:
    * 트리플렛 손실(Triplet Loss) 또는 콘트라스트 손실(Contrastive Loss)을 사용하여 임베딩 공간을 학습합니다.
    * 학습된 임베딩 공간에서 동일 클래스와 다른 클래스의 유사도 분포를 평가합니다.))

오늘 1에서 8까지 총 8가지 class similarity 분석 진행했습니다.
(오늘은 저번에 주셨던 data의 모든 클래스 사용하여 실행했습니다!
용량이 커서 모든 데이터를 사용하진 않고 각 클래스별로 이미지 파일은 조금씩 넣어놓고 돌려봤어요)

4번은 1번과 같은 계산 방식이라 넘기셔도 되고,
6, 7, 8번은 클래스 유사성 계산 방식이 아니고 모델 평가 지표였습니다. 
참고 부탁드립니다!

각각에 대한 설명, 코드와 결과는 simResult.ipynb에 정리해놓았습니다!

**Comment by Hyeonjha** [comment]
Jun 7

Original comment by JihoonPark93

자세하게 잘 적어주셨네요! 👍 마지막 8번 방법 결과 데이터를 normalize 해서 결과를 비교해보면 더 좋을 것 같아요! 각 모델별 X 축 범위도 모두 똑같이 설정해서 보면 더 좋을 것 같아요
전체적으로 어느 방법을 쓰든 모델 평가 결과가 유사하게 나와서 모든 방법을 사용할 필요는 없을 것 같고, 앞으로 추가 모델 평가가 필요한 경우 6, 8번으로 평가를 진행하면 좋을 것 같아요. 6번의 경우 일반적으로 모델 평가에 많이 사용하는 방식이고, 8번의 경우 실제 데이터셋에서 분포가 어떻게 되는지를 시각적으로 바로 확인할 수 있어 좋아보입니다
두 평가 방법에 대해서 code refactoring 좀 더 진행하고 다음 업무로 넘어가도 될 것 같아요

다양한 평가 방법으로 다 비교했을 때 ResNet18 이 연산 시간이나 결과 면에서 괜찮아보이긴 하지만, 괜찮은 평가 방식을 찾았으니 조금 더 최신 모델을 적용해서 평가해봐도 좋을 것 같아요 
아래 sota 모델들 중 model size, score 를 참고해서 + 코드 찾기도 편한걸로 상위 rank 몇개만 더 테스트해보고 issue close 해도 될 것 같습니다!
https://paperswithcode.com/sota/image-classification-on-imagenet

**Comment by Hyeonjha** [comment]
Jun 10

오늘 진행한 내용은 simResult_.ipynb에 정리했습니다!

말씀하신대로 데이터 normalization, X 축 범위 통일시켜 6번, 8번 사용하여 모델 평가 시도해봤습니다.
최신 모델 사용하기 위해 코드 찾고 실행해보다가 시간을 좀 많이 썼네요,,!
최신 모델 새로 구현하여 적용한 성능 평가는 제가 더 수정하고 다뤄봐야 할 것 같아서 내일 마무리 하도록 하겠습니다!

**Comment by Hyeonjha** [comment]
Jun 11

지훈님 오늘 진행한 내용
ImageImbedding/simResult_Final.ipynb 이 파일에 정리해놨습니다!
(https://github.com/JihoonPark93/ImageImbedding/blob/HJ/simResult_Final.ipynb)
확인 부탁드립니다:))
